Add support for PHP 8 and phpunit/php-code-coverage 9

jaylinski opened the pull request:
* Added support for PHP 8 (c5a507b)
* Added support for phpunit/php-code-coverage 9 (8383dce, #36)

jaylinski commented on src/CodeCoverageExtension.php:
```diff
@@ -50,17 +49,13 @@ public function load(ServiceContainer $container, array $params = []): void
             }
 
-            return $coverage;
+            return new CodeCoverage($driver, $container->get('code_coverage.filter'));
```
PHPStan complains, because it doesn't get the version check... I'm not sure how to work around this. @drupol Any ideas?

drupol replied:
I'll try to give it a try soon!

drupol replied:
Here's my [log](https://del.dog/icupelivyr.go)... investigating.

drupol replied:
Ok I see.

CodeCoverage's first parameter is not supposed to be null, but in the code you wrote, it is possible that `$driver` gets `null`.

How about doing this?

```php
        $container->define('code_coverage', static function ($container) {
            if (version_compare(Version::id(), '9.0.0') < 0) {
                throw new Exception(
                    'There is no available coverage driver to be used.'
                );
            }

            /** @var Filter $filter */
            $filter = $container->get('code_coverage.filter');

            return new CodeCoverage(
                (new Selector())->forLineCoverage($container->get('code_coverage.filter')),
                $filter
            );
        });
```

shulard replied:
Hum, with this code, the coverage will not be possible in version under 9.0.0. I think that we need to maintain the `null` value.

I'll go for something like this : 

```php
        $container->define('code_coverage', static function ($container) {
            /** @var Filter $filter */
            $filter = $container->get('code_coverage.filter');

            if (version_compare(Version::id(), '9.0.0') < 0) {
                // This is valid in version under 9 and let the package choose the correct tool
                return new CodeCoverage(null, $filter);
            } else {
                return new CodeCoverage(
                    (new Selector())->forLineCoverage($container->get('code_coverage.filter')),
                    $filter
                );
        });
```

Also, I think that catching the exception and throwing our own is better from a DX perspective because it'll help knowing where the issue comme from…

drupol replied:
Oh... if this is valid, then this is indeed the proper way to go. We could also eventually get rid of the else clause.

```php
$container->define('code_coverage', static function ($container) {
            /** @var Filter $filter */
            $filter = $container->get('code_coverage.filter');

            if (version_compare(Version::id(), '9.0.0') < 0) {
                // This is valid in version under 9 and let the package choose the correct tool
                return new CodeCoverage(null, $filter);
            }

                return new CodeCoverage(
                    (new Selector())->forLineCoverage($container->get('code_coverage.filter')),
                    $filter
                );
        });
```

jaylinski replied:
I added this, but even with `// @phpstan-ignore-line` I'm getting errors...

drupol commented on composer.json:
```diff
@@ -38,9 +38,9 @@
     ],
     "require": {
-        "php": ">= 7.2",
+        "php": "^7.2 || ^8.0",
```
Why this change when the Symfony team doesn't recommend this?

shulard replied:
Do you know why they don't recommend this ?

jaylinski replied:
Yeah, that makes sense. I thought `>= 7.2` wouldn't allow for PHP 8, but it obviously does. Will change it back.

drupol reviewed (changes requested):
The CodeCoverageExtension class's load method should strictly follow its interface implementation and should be changed to:

`public function load(ServiceContainer $container, array $params): void`

Tests will have to be updated, but it's quite easy, just adding `[]` here and there.

drupol commented on spec/Listener/CodeCoverageListenerSpec.php:
```diff
@@ -5,15 +5,15 @@
+ * final and we cannot mock them. The tests should be converted into proper
+ * functional (integration) tests instead. This file is left for reference.
+ *
+ * @see https://github.com/leanphp/phpspec-code-coverage/issues/19
```
How about adding an interface to those final classes? This would fix all the issues.

drupol replied:
Adding the interface is not going to be an option, see the comment here: https://github.com/sebastianbergmann/php-code-coverage/pull/833

jaylinski replied:
I would prefer to add this in separate commit.

drupol replied:
I think we actually should not do that. We most probably should stub the Driver, just like the PHPUnit maintainer explained in its comment.

shulard commented on src/Exception/NoCoverageDriverAvailableException.php:
```diff
@@ -1,27 +0,0 @@
- *
- * @author Stéphane Hulard
- */
-class NoCoverageDriverAvailableException extends RuntimeException
```
Why removing this specific exception ?

jaylinski replied:
Because we just rethrow the same exception from here: https://github.com/sebastianbergmann/php-code-coverage/blob/9.2.3/src/Exception/NoCodeCoverageDriverAvailableException.php

I think that's unnecessary. Should I add it again?

shulard replied:
This is a case we know and throwing a custom exception help finding where is the issue point instead requiring to check another external lib code...

I use this kind of technique in some projects and find it helpful because it help :

* knowing the possible errors, I mean the one understood in the code logic and that are « normal »,
* make a difference with the others that are not valid from our library perspective.

But it’s my POV, I totally understand your point too 😄.

@drupol what do you think ?

jaylinski replied:
@shulard I added the try-catch again. :+1:

jaylinski commented:
@drupol Apart from the PHPStan issue, there is also this error on PHP 8:

```
Problem 1
    - drupol/php-conventions[1.7.4, ..., 1.8.19] require friendsoftwig/twigcs ^3.2.2 -> satisfiable by friendsoftwig/twigcs[v3.2.2].
    - friendsoftwig/twigcs v3.2.2 requires php ^7.0 -> your php version (8.0.0) does not satisfy that requirement.
    - Root composer.json requires drupol/php-conventions ^1.7.1 || ^1.8.16 -> satisfiable by drupol/php-conventions[1.7.4, ..., 1.8.19].
```

Can you have a look at this?

jaylinski commented:
@shulard @drupol After lots of back and forwards, I changed my approach. I think it is best for us to release (another) major version (v6) targeting PHP >= 7.3 and phpunit-code-coverage v9.

This way we can hand of the backwards-compatibility check to Composer. Composer will choose our v5 version if PHP 7.2 support or phpunit-code-coverage < v9 support is required.
If the user is using PHP 7.3 and has already upgraded to phpunit v9, it will use our v6 version.

We still can maintain version 5 for a year, until PHP 7.2 support is not supported anymore (https://www.php.net/supported-versions.php).

I also added a compatibility-table to our readme to clarify our dependencies.

I disabled GrumPHP on PHP 8.0 for now, since `drupol/php-conventions`, or more precisely, it's dependencies (php-cs-fixer, twig-linter, ...), don't support PHP 8.0 yet. The CI checks will still run for PHP 7.3 and PHP 7.4 though.

drupol commented:
I prefer this approach indeed :-)

shulard commented:
I prefer this approach too, this is a nice way to handle those updates and move forward with the ecosystem !

shulard commented:
I just created a new branch `releases/5.0` that will be used as the maintenance branch for the 5.x version of this package. I also added some rules in the project settings to avoid deleting this branch ^^.

Thank you @jaylinski for this awesome PR (and I love the version table added in the README). I merge it, do you think we can already create the tag and publish the release ?

jaylinski commented:
@shulard I already tested my branch in a PHP 8.0 container. The coverage worked fine. So I think we can create a tag.

shulard commented:
Super, the `v6.0.0` version was just released !
